File: android/app/src/main/java/com/vxin/app/core/call/GroupCallManager.kt
package com.vxin.app.core.call

import android.content.Context
import android.util.Log
import com.vxin.app.core.auth.SessionManager
import com.vxin.app.core.di.AppScope
import com.vxin.app.core.realtime.SocketManager
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import org.webrtc.AudioSource
import org.webrtc.AudioTrack
import org.webrtc.Camera2Enumerator
import org.webrtc.CameraVideoCapturer
import org.webrtc.DefaultVideoDecoderFactory
import org.webrtc.DefaultVideoEncoderFactory
import org.webrtc.EglBase
import org.webrtc.IceCandidate
import org.webrtc.MediaConstraints
import org.webrtc.MediaStream
import org.webrtc.PeerConnection
import org.webrtc.PeerConnectionFactory
import org.webrtc.RtpReceiver
import org.webrtc.SessionDescription
import org.webrtc.SurfaceTextureHelper
import org.webrtc.VideoCapturer
import org.webrtc.VideoSource
import org.webrtc.VideoTrack
import javax.inject.Inject
import javax.inject.Singleton

enum class GroupCallStage { IDLE, CONNECTING, CONNECTED, ENDED }

data class GroupCallState(
    val stage: GroupCallStage = GroupCallStage.IDLE,
    val callId: String = "",
    val conversationId: String = "",
    val isVideo: Boolean = false,
    val participants: List<String> = emptyList(), // 远端成员 id（不含自己）
    val micEnabled: Boolean = true,
    val cameraEnabled: Boolean = true,
)

/**
 * 群音视频通话（mesh）。信令协议见 backend-v2/docs/GROUP_CALL.md。
 * 与 [CallManager] 各自独立；本地音视频轨只建一份，加入到每条 PeerConnection。
 * 防 glare：新加入者只 answer；既有成员收到 peer_joined 才向其 createOffer。
 */
@Singleton
class GroupCallManager @Inject constructor(
    @ApplicationContext private val context: Context,
    private val socketManager: SocketManager,
    private val sessionManager: SessionManager,
    private val turnApi: com.vxin.app.data.api.TurnApi,
    @AppScope private val scope: CoroutineScope,
) {
    val eglBase: EglBase = EglBase.create()

    private var factory: PeerConnectionFactory? = null
    private var audioSource: AudioSource? = null
    private var localAudioTrack: AudioTrack? = null
    private var videoSource: VideoSource? = null
    private var videoCapturer: VideoCapturer? = null
    private var surfaceHelper: SurfaceTextureHelper? = null
    var localVideoTrack: VideoTrack? = null
        private set

    private data class Peer(
        val pc: PeerConnection,
        var remoteDescSet: Boolean = false,
        val pendingIce: MutableList<IceCandidate> = mutableListOf(),
    )
    private val peers = LinkedHashMap<String, Peer>()

    private val _state = MutableStateFlow(GroupCallState())
    val state: StateFlow<GroupCallState> = _state.asStateFlow()

    // 远端视频轨：peerId -> VideoTrack，供 UI 宫格渲染
    private val _remoteTracks = MutableStateFlow<Map<String, VideoTrack>>(emptyMap())
    val remoteTracks: StateFlow<Map<String, VideoTrack>> = _remoteTracks.asStateFlow()

    private val fallbackIceServers = listOf(
        PeerConnection.IceServer.builder("stun:stun.l.google.com:19302").createIceServer(),
    )
    @Volatile private var iceServers: List<PeerConnection.IceServer> = fallbackIceServers

    init {
        ensureFactory()
        observeSignaling()
    }

    private fun ensureFactory() {
        if (factory != null) return
        PeerConnectionFactory.initialize(
            PeerConnectionFactory.InitializationOptions.builder(context).createInitializationOptions()
        )
        factory = PeerConnectionFactory.builder()
            .setVideoEncoderFactory(DefaultVideoEncoderFactory(eglBase.eglBaseContext, true, true))
            .setVideoDecoderFactory(DefaultVideoDecoderFactory(eglBase.eglBaseContext))
            .createPeerConnectionFactory()
    }

    private suspend fun refreshIceServers() {
        try {
            val creds = turnApi.getCredentials()
            val servers = creds.iceServers.mapNotNull { dto ->
                if (dto.urls.isEmpty()) return@mapNotNull null
                PeerConnection.IceServer.builder(dto.urls).apply {
                    dto.username?.let { setUsername(it) }
                    dto.credential?.let { setPassword(it) }
                }.createIceServer()
            }
            if (servers.isNotEmpty()) iceServers = servers
        } catch (e: Exception) {
            Log.w(TAG, "refreshIceServers failed, fallback STUN", e)
        }
    }

    // ── 对外动作 ───────────────────────────────────────────
    /** 发起群通话 */
    fun start(conversationId: String, video: Boolean) {
        if (_state.value.stage != GroupCallStage.IDLE && _state.value.stage != GroupCallStage.ENDED) return
        _state.value = GroupCallState(GroupCallStage.CONNECTING, conversationId = conversationId, isVideo = video)
        scope.launch {
            refreshIceServers()
            if (_state.value.stage == GroupCallStage.ENDED) return@launch
            createLocalMedia(video)
            socketManager.emitGroupCallStart(conversationId, if (video) "video" else "audio")
        }
    }

    /** 加入已有群通话 */
    fun join(callId: String, conversationId: String, video: Boolean) {
        if (_state.value.stage != GroupCallStage.IDLE && _state.value.stage != GroupCallStage.ENDED) return
        _state.value = GroupCallState(GroupCallStage.CONNECTING, callId, conversationId, isVideo = video)
        scope.launch {
            refreshIceServers()
            if (_state.value.stage == GroupCallStage.ENDED) return@launch
            createLocalMedia(video)
            socketManager.emitGroupCallJoin(callId)
        }
    }

    /** 挂断/离开 */
    fun hangup() {
        val cid = _state.value.callId
        if (cid.isNotEmpty()) socketManager.emitGroupCallLeave(cid)
        cleanup()
    }

    fun toggleMic() {
        val enabled = !_state.value.micEnabled
        localAudioTrack?.setEnabled(enabled)
        _state.update { it.copy(micEnabled = enabled) }
    }

    fun toggleCamera() {
        val enabled = !_state.value.cameraEnabled
        localVideoTrack?.setEnabled(enabled)
        _state.update { it.copy(cameraEnabled = enabled) }
    }

    fun switchCamera() { (videoCapturer as? CameraVideoCapturer)?.switchCamera(null) }

    fun consumeEnded() {
        if (_state.value.stage == GroupCallStage.ENDED) _state.value = GroupCallState()
    }

    // ── 信令处理 ───────────────────────────────────────────
    private fun observeSignaling() {
        scope.launch {
            socketManager.groupCallStartedEvents.collect { e ->
                if (_state.value.stage == GroupCallStage.ENDED) return@collect
                _state.update { it.copy(stage = GroupCallStage.CONNECTED, callId = e.callId) }
            }
        }
        scope.launch {
            socketManager.groupCallPeersEvents.collect { e ->
                if (_state.value.callId.isNotEmpty() && e.callId != _state.value.callId) return@collect
                _state.update { it.copy(stage = GroupCallStage.CONNECTED, callId = e.callId) }
                // 作为 answerer：为既有成员预建 PC，等其 offer
                e.peers.forEach { pid -> peerFor(pid) }
                _state.update { it.copy(participants = peers.keys.toList()) }
            }
        }
        scope.launch {
            socketManager.groupCallPeerJoinedEvents.collect { e ->
                if (e.callId != _state.value.callId) return@collect
                val peer = peerFor(e.userId)
                _state.update { it.copy(participants = peers.keys.toList()) }
                // 既有成员向新 peer 发 offer
                peer.pc.createOffer(object : SimpleSdpObserver() {
                    override fun onCreateSuccess(desc: SessionDescription) {
                        peer.pc.setLocalDescription(SimpleSdpObserver(), desc)
                        socketManager.emitGroupCallOffer(_state.value.callId, e.userId, desc.description)
                    }
                }, mediaConstraints())
            }
        }
        scope.launch {
            socketManager.groupCallOfferEvents.collect { e ->
                if (e.callId != _state.value.callId) return@collect
                val peer = peerFor(e.from)
                _state.update { it.copy(participants = peers.keys.toList()) }
                peer.pc.setRemoteDescription(object : SimpleSdpObserver() {
                    override fun onSetSuccess() {
                        peer.remoteDescSet = true; drainIce(e.from)
                        peer.pc.createAnswer(object : SimpleSdpObserver() {
                            override fun onCreateSuccess(desc: SessionDescription) {
                                peer.pc.setLocalDescription(SimpleSdpObserver(), desc)
                                socketManager.emitGroupCallAnswer(_state.value.callId, e.from, desc.description)
                            }
                        }, mediaConstraints())
                    }
                }, SessionDescription(SessionDescription.Type.OFFER, e.sdp))
            }
        }
        scope.launch {
            socketManager.groupCallAnswerEvents.collect { e ->
                val peer = peers[e.from] ?: return@collect
                peer.pc.setRemoteDescription(object : SimpleSdpObserver() {
                    override fun onSetSuccess() { peer.remoteDescSet = true; drainIce(e.from) }
                }, SessionDescription(SessionDescription.Type.ANSWER, e.sdp))
            }
        }
        scope.launch {
            socketManager.groupCallIceEvents.collect { e ->
                val peer = peers[e.from] ?: return@collect
                val cand = IceCandidate(e.sdpMid, e.sdpMLineIndex, e.candidate)
                if (peer.remoteDescSet) peer.pc.addIceCandidate(cand) else peer.pendingIce.add(cand)
            }
        }
        scope.launch {
            socketManager.groupCallPeerLeftEvents.collect { e -> removePeer(e.userId) }
        }
        scope.launch {
            socketManager.groupCallErrorEvents.collect { e ->
                Log.w(TAG, "group call error: ${e.reason}")
                if (_state.value.stage != GroupCallStage.CONNECTED) cleanup()
            }
        }
    }

    private fun drainIce(peerId: String) {
        val peer = peers[peerId] ?: return
        peer.pendingIce.forEach { peer.pc.addIceCandidate(it) }
        peer.pendingIce.clear()
    }

    // ── WebRTC ───────────────────────────────────────────
    private fun createLocalMedia(video: Boolean) {
        val f = factory ?: return
        audioSource = f.createAudioSource(MediaConstraints())
        localAudioTrack = f.createAudioTrack("g_audio", audioSource).apply { setEnabled(true) }
        if (video) {
            val capturer = createCameraCapturer() ?: return
            videoCapturer = capturer
            surfaceHelper = SurfaceTextureHelper.create("GCCapture", eglBase.eglBaseContext)
            videoSource = f.createVideoSource(false)
            capturer.initialize(surfaceHelper, context, videoSource!!.capturerObserver)
            runCatching { capturer.startCapture(1280, 720, 30) }
            localVideoTrack = f.createVideoTrack("g_video", videoSource).apply { setEnabled(true) }
        }
    }

    private fun createCameraCapturer(): VideoCapturer? {
        val enumerator = Camera2Enumerator(context)
        val names = enumerator.deviceNames
        names.firstOrNull { enumerator.isFrontFacing(it) }?.let { return enumerator.createCapturer(it, null) }
        names.firstOrNull()?.let { return enumerator.createCapturer(it, null) }
        return null
    }

    // 为某 peer 建立 PeerConnection（含本地轨）。幂等。
    private fun peerFor(peerId: String): Peer {
        peers[peerId]?.let { return it }
        val f = factory!!
        val config = PeerConnection.RTCConfiguration(iceServers).apply {
            sdpSemantics = PeerConnection.SdpSemantics.UNIFIED_PLAN
        }
        val pc = f.createPeerConnection(config, object : PeerConnection.Observer {
            override fun onIceCandidate(candidate: IceCandidate) {
                socketManager.emitGroupCallIce(_state.value.callId, peerId, candidate.sdp, candidate.sdpMid, candidate.sdpMLineIndex)
            }
            override fun onAddTrack(receiver: RtpReceiver, streams: Array<out MediaStream>?) {
                (receiver.track() as? VideoTrack)?.let { vt ->
                    _remoteTracks.update { it + (peerId to vt) }
                }
            }
            override fun onIceConnectionChange(state: PeerConnection.IceConnectionState) {
                if (state == PeerConnection.IceConnectionState.FAILED ||
                    state == PeerConnection.IceConnectionState.CLOSED) removePeer(peerId)
            }
            override fun onSignalingChange(p0: PeerConnection.SignalingState?) {}
            override fun onIceConnectionReceivingChange(p0: Boolean) {}
            override fun onIceGatheringChange(p0: PeerConnection.IceGatheringState?) {}
            override fun onIceCandidatesRemoved(p0: Array<out IceCandidate>?) {}
            override fun onAddStream(p0: MediaStream?) {}
            override fun onRemoveStream(p0: MediaStream?) {}
            override fun onDataChannel(p0: org.webrtc.DataChannel?) {}
            override fun onRenegotiationNeeded() {}
        })!!
        localAudioTrack?.let { pc.addTrack(it, listOf(STREAM_ID)) }
        localVideoTrack?.let { pc.addTrack(it, listOf(STREAM_ID)) }
        val peer = Peer(pc)
        peers[peerId] = peer
        return peer
    }

    private fun removePeer(peerId: String) {
        peers.remove(peerId)?.let { runCatching { it.pc.close(); it.pc.dispose() } }
        _remoteTracks.update { it - peerId }
        _state.update { it.copy(participants = peers.keys.toList()) }
    }

    private fun mediaConstraints() = MediaConstraints().apply {
        mandatory.add(MediaConstraints.KeyValuePair("OfferToReceiveAudio", "true"))
        mandatory.add(MediaConstraints.KeyValuePair("OfferToReceiveVideo", if (_state.value.isVideo) "true" else "false"))
    }

    private fun cleanup() {
        peers.values.forEach { runCatching { it.pc.close(); it.pc.dispose() } }
        peers.clear()
        _remoteTracks.value = emptyMap()
        runCatching { videoCapturer?.stopCapture() }
        runCatching { videoCapturer?.dispose() }; videoCapturer = null
        surfaceHelper?.dispose(); surfaceHelper = null
        localVideoTrack = null
        runCatching { videoSource?.dispose() }; videoSource = null
        runCatching { audioSource?.dispose() }; audioSource = null
        localAudioTrack = null
        _state.value = _state.value.copy(stage = GroupCallStage.ENDED, participants = emptyList())
    }

    private companion object {
        const val STREAM_ID = "g_stream"
        const val TAG = "GroupCallManager"
    }
}

File: ios/Vxin/Core/Call/GroupCallManager.swift
import Foundation
import Combine
import AVFoundation
import WebRTC

enum GroupCallStage { case idle, connecting, connected, ended }

struct GroupCallInvite: Identifiable {
    let callId: String
    let conversationId: String
    let type: String
    let from: String
    let fromName: String
    var id: String { callId }
}

struct GroupCallState {
    var stage: GroupCallStage = .idle
    var callId: String = ""
    var conversationId: String = ""
    var isVideo: Bool = false
    var participants: [String] = []   // 远端成员 id（不含自己）
    var micEnabled: Bool = true
    var cameraEnabled: Bool = true
}

/// 群音视频通话（mesh）。信令协议见 backend-v2/docs/GROUP_CALL.md。
/// 与 [CallManager] 各自独立；本地音视频轨只建一份，加入到每条 PeerConnection。
/// 防 glare：新加入者只 answer；既有成员收到 peer_joined 才向其 createOffer。
final class GroupCallManager: NSObject, ObservableObject {
    static let shared = GroupCallManager()

    @Published private(set) var state = GroupCallState()
    @Published private(set) var remoteTracks: [String: RTCVideoTrack] = [:]
    @Published var pendingInvite: GroupCallInvite?

    private let factory: RTCPeerConnectionFactory
    private var localAudioTrack: RTCAudioTrack?
    private(set) var localVideoTrack: RTCVideoTrack?
    private var videoCapturer: RTCCameraVideoCapturer?

    final class PeerEntry {
        let pc: RTCPeerConnection
        let delegate: GCPeerDelegate
        var remoteDescSet = false
        var pendingIce: [RTCIceCandidate] = []
        init(pc: RTCPeerConnection, delegate: GCPeerDelegate) { self.pc = pc; self.delegate = delegate }
    }
    private var peers: [String: PeerEntry] = [:]

    private var iceServers = [RTCIceServer(urlStrings: ["stun:stun.l.google.com:19302"])]
    private var cancellables = Set<AnyCancellable>()
    private let socket = SocketService.shared

    private override init() {
        RTCInitializeSSL()
        factory = RTCPeerConnectionFactory(
            encoderFactory: RTCDefaultVideoEncoderFactory(),
            decoderFactory: RTCDefaultVideoDecoderFactory()
        )
        super.init()
        observeSignaling()
    }

    func activate() {}

    // MARK: - 音频会话（WebRTC）
    /// 建流前配置 RTCAudioSession 为通话模式(.playAndRecord/.voiceChat)。
    /// 走 RTCAudioSession 而非裸 AVAudioSession：WebRTC 内部持有会话，只有经其配置才与音频单元协调。
    /// 通话期间语音消息播放(AudioPlayerService)不应抢占本会话。
    private func configureAudioSession() {
        let session = RTCAudioSession.sharedInstance()
        session.lockForConfiguration()
        do {
            try session.setCategory(
                AVAudioSession.Category.playAndRecord.rawValue,
                with: [.allowBluetooth]
            )
            try session.setMode(AVAudioSession.Mode.voiceChat.rawValue)
            try session.setActive(true)
        } catch {
            // 配置失败不阻断通话；WebRTC 兜底默认会话
        }
        session.unlockForConfiguration()
    }

    /// 通话结束释放音频会话，交还系统。
    private func deactivateAudioSession() {
        let session = RTCAudioSession.sharedInstance()
        session.lockForConfiguration()
        try? session.setActive(false)
        session.unlockForConfiguration()
    }

    private func refreshIceServers() async {
        do {
            let creds: TurnCredentials = try await APIClient.shared.send("api/turn/credentials")
            let servers = creds.iceServers.compactMap { dto -> RTCIceServer? in
                guard !dto.urls.isEmpty else { return nil }
                if let u = dto.username, let c = dto.credential {
                    return RTCIceServer(urlStrings: dto.urls, username: u, credential: c)
                }
                return RTCIceServer(urlStrings: dto.urls)
            }
            if !servers.isEmpty { iceServers = servers }
        } catch { /* 兜底 STUN */ }
    }

    // MARK: - 对外动作
    func start(conversationId: String, video: Bool) {
        guard state.stage == .idle || state.stage == .ended else { return }
        pendingInvite = nil
        state = GroupCallState(stage: .connecting, conversationId: conversationId, isVideo: video)
        Task { @MainActor in
            await refreshIceServers()
            guard state.stage != .ended else { return }
            configureAudioSession()             // 建流前配好通话音频会话
            createLocalMedia(video: video)
            socket.emitGroupCallStart(conversationId: conversationId, type: video ? "video" : "audio")
        }
    }

    func join(callId: String, conversationId: String, video: Bool) {
        guard state.stage == .idle || state.stage == .ended else { return }
        pendingInvite = nil
        state = GroupCallState(stage: .connecting, callId: callId, conversationId: conversationId, isVideo: video)
        Task { @MainActor in
            await refreshIceServers()
            guard state.stage != .ended else { return }
            configureAudioSession()             // 建流前配好通话音频会话
            createLocalMedia(video: video)
            socket.emitGroupCallJoin(callId: callId)
        }
    }

    func hangup() {
        if !state.callId.isEmpty { socket.emitGroupCallLeave(callId: state.callId) }
        cleanup()
    }

    func toggleMic() {
        let on = !state.micEnabled
        localAudioTrack?.isEnabled = on
        state.micEnabled = on
    }
    func toggleCamera() {
        let on = !state.cameraEnabled
        localVideoTrack?.isEnabled = on
        state.cameraEnabled = on
    }
    func switchCamera() {
        guard let capturer = videoCapturer else { return }
        let current = capturer.captureSession.inputs.compactMap { ($0 as? AVCaptureDeviceInput)?.device.position }.first ?? .front
        startCapture(position: current == .front ? .back : .front)
    }
    func consumeEnded() { if state.stage == .ended { state = GroupCallState() } }

    // MARK: - 信令
    private func observeSignaling() {
        socket.gcInvite.receive(on: DispatchQueue.main).sink { [weak self] inv in
            guard let self else { return }
            if self.state.stage == .connecting || self.state.stage == .connected { return }
            self.pendingInvite = GroupCallInvite(callId: inv.callId, conversationId: inv.conversationId, type: inv.type, from: inv.from, fromName: inv.fromName)
        }.store(in: &cancellables)

        socket.gcStarted.receive(on: DispatchQueue.main).sink { [weak self] (callId, _) in
            guard let self, self.state.stage != .ended else { return }
            self.state.stage = .connected; self.state.callId = callId
        }.store(in: &cancellables)

        socket.gcPeers.receive(on: DispatchQueue.main).sink { [weak self] (callId, _, peers) in
            guard let self else { return }
            if !self.state.callId.isEmpty && callId != self.state.callId { return }
            self.state.stage = .connected; self.state.callId = callId
            peers.forEach { _ = self.peerFor($0) }   // answerer：预建 PC 等 offer
            self.state.participants = Array(self.peers.keys)
        }.store(in: &cancellables)

        socket.gcPeerJoined.receive(on: DispatchQueue.main).sink { [weak self] (callId, userId) in
            guard let self, callId == self.state.callId, let entry = self.peerFor(userId) else { return }
            self.state.participants = Array(self.peers.keys)
            entry.pc.offer(for: self.mediaConstraints()) { [weak self] desc, err in
                guard let self, let desc, err == nil else { return }
                entry.pc.setLocalDescription(desc) { _ in }
                self.socket.emitGroupCallOffer(callId: self.state.callId, to: userId, sdp: desc.sdp)
            }
        }.store(in: &cancellables)

        socket.gcOffer.receive(on: DispatchQueue.main).sink { [weak self] (callId, from, sdp) in
            guard let self, callId == self.state.callId, let entry = self.peerFor(from) else { return }
            self.state.participants = Array(self.peers.keys)
            entry.pc.setRemoteDescription(RTCSessionDescription(type: .offer, sdp: sdp)) { [weak self] err in
                guard let self, err == nil else { return }
                entry.remoteDescSet = true; self.drainIce(from)
                entry.pc.answer(for: self.mediaConstraints()) { [weak self] desc, err in
                    guard let self, let desc, err == nil else { return }
                    entry.pc.setLocalDescription(desc) { _ in }
                    self.socket.emitGroupCallAnswer(callId: self.state.callId, to: from, sdp: desc.sdp)
                }
            }
        }.store(in: &cancellables)

        socket.gcAnswer.receive(on: DispatchQueue.main).sink { [weak self] (_, from, sdp) in
            guard let self, let entry = self.peers[from] else { return }
            entry.pc.setRemoteDescription(RTCSessionDescription(type: .answer, sdp: sdp)) { [weak self] err in
                guard let self, err == nil else { return }
                entry.remoteDescSet = true; self.drainIce(from)
            }
        }.store(in: &cancellables)

        socket.gcIce.receive(on: DispatchQueue.main).sink { [weak self] (_, from, candidate, sdpMid, idx) in
            guard let self, let entry = self.peers[from] else { return }
            let cand = RTCIceCandidate(sdp: candidate, sdpMLineIndex: idx, sdpMid: sdpMid)
            if entry.remoteDescSet { entry.pc.add(cand) } else { entry.pendingIce.append(cand) }
        }.store(in: &cancellables)

        socket.gcPeerLeft.receive(on: DispatchQueue.main).sink { [weak self] (_, userId) in
            self?.removePeer(userId)
        }.store(in: &cancellables)

        socket.gcError.receive(on: DispatchQueue.main).sink { [weak self] _ in
            guard let self else { return }
            if self.state.stage != .connected { self.cleanup() }
        }.store(in: &cancellables)
    }

    private func drainIce(_ peerId: String) {
        guard let entry = peers[peerId] else { return }
        entry.pendingIce.forEach { entry.pc.add($0) }
        entry.pendingIce.removeAll()
    }

    // MARK: - per-peer 回调（由 GCPeerDelegate 转发）
    func onIce(_ peerId: String, _ candidate: RTCIceCandidate) {
        socket.emitGroupCallIce(callId: state.callId, to: peerId, candidate: candidate.sdp, sdpMid: candidate.sdpMid, sdpMLineIndex: candidate.sdpMLineIndex)
    }
    func onRemoteVideo(_ peerId: String, _ track: RTCVideoTrack) {
        DispatchQueue.main.async { self.remoteTracks[peerId] = track }
    }
    func onIceState(_ peerId: String, _ newState: RTCIceConnectionState) {
        if newState == .failed || newState == .closed {
            DispatchQueue.main.async { self.removePeer(peerId) }
        }
    }

    // MARK: - WebRTC
    private func createLocalMedia(video: Bool) {
        let audioSource = factory.audioSource(with: RTCMediaConstraints(mandatoryConstraints: nil, optionalConstraints: nil))
        localAudioTrack = factory.audioTrack(with: audioSource, trackId: "g_audio")
        if video {
            let videoSource = factory.videoSource()
            videoCapturer = RTCCameraVideoCapturer(delegate: videoSource)
            localVideoTrack = factory.videoTrack(with: videoSource, trackId: "g_video")
            startCapture(position: .front)
        }
    }

    private func startCapture(position: AVCaptureDevice.Position) {
        guard let capturer = videoCapturer else { return }
        let devices = RTCCameraVideoCapturer.captureDevices()
        guard let device = devices.first(where: { $0.position == position }) ?? devices.first else { return }
        let formats = RTCCameraVideoCapturer.supportedFormats(for: device)
        let format = formats.sorted {
            let d1 = CMVideoFormatDescriptionGetDimensions($0.formatDescription)
            let d2 = CMVideoFormatDescriptionGetDimensions($1.formatDescription)
            return d1.width * d1.height < d2.width * d2.height
        }.first(where: { CMVideoFormatDescriptionGetDimensions($0.formatDescription).width >= 640 }) ?? formats.last
        guard let format else { return }
        let fps = format.videoSupportedFrameRateRanges.map { $0.maxFrameRate }.max() ?? 30
        capturer.startCapture(with: device, format: format, fps: Int(min(fps, 30)))
    }

    private func peerFor(_ peerId: String) -> PeerEntry? {
        if let e = peers[peerId] { return e }
        let config = RTCConfiguration()
        config.iceServers = iceServers
        config.sdpSemantics = .unifiedPlan
        let delegate = GCPeerDelegate(peerId: peerId, manager: self)
        let constraints = RTCMediaConstraints(mandatoryConstraints: nil, optionalConstraints: nil)
        guard let pc = factory.peerConnection(with: config, constraints: constraints, delegate: delegate) else { return nil }
        if let a = localAudioTrack { pc.add(a, streamIds: ["g_stream"]) }
        if let v = localVideoTrack { pc.add(v, streamIds: ["g_stream"]) }
        let entry = PeerEntry(pc: pc, delegate: delegate)
        peers[peerId] = entry
        return entry
    }

    private func removePeer(_ peerId: String) {
        peers[peerId]?.pc.close()
        peers[peerId] = nil
        remoteTracks[peerId] = nil
        state.participants = Array(peers.keys)
    }

    private func mediaConstraints() -> RTCMediaConstraints {
        RTCMediaConstraints(
            mandatoryConstraints: ["OfferToReceiveAudio": "true", "OfferToReceiveVideo": state.isVideo ? "true" : "false"],
            optionalConstraints: nil
        )
    }

    private func cleanup() {
        peers.values.forEach { $0.pc.close() }
        peers.removeAll()
        remoteTracks.removeAll()
        videoCapturer?.stopCapture()
        videoCapturer = nil
        localVideoTrack = nil
        localAudioTrack = nil
        deactivateAudioSession()            // 释放通话音频会话
        state.stage = .ended
        state.participants = []
    }
}

/// 每条 PeerConnection 一个委托，把回调连同 peerId 转回 manager。
final class GCPeerDelegate: NSObject, RTCPeerConnectionDelegate {
    let peerId: String
    weak var manager: GroupCallManager?
    init(peerId: String, manager: GroupCallManager) { self.peerId = peerId; self.manager = manager }

    func peerConnection(_ pc: RTCPeerConnection, didGenerate candidate: RTCIceCandidate) {
        manager?.onIce(peerId, candidate)
    }
    func peerConnection(_ pc: RTCPeerConnection, didAdd rtpReceiver: RTCRtpReceiver, streams mediaStreams: [RTCMediaStream]) {
        if let track = rtpReceiver.track as? RTCVideoTrack { manager?.onRemoteVideo(peerId, track) }
    }
    func peerConnection(_ pc: RTCPeerConnection, didChange newState: RTCIceConnectionState) {
        manager?.onIceState(peerId, newState)
    }
    func peerConnection(_ pc: RTCPeerConnection, didChange stateChanged: RTCSignalingState) {}
    func peerConnection(_ pc: RTCPeerConnection, didAdd stream: RTCMediaStream) {}
    func peerConnection(_ pc: RTCPeerConnection, didRemove stream: RTCMediaStream) {}
    func peerConnectionShouldNegotiate(_ pc: RTCPeerConnection) {}
    func peerConnection(_ pc: RTCPeerConnection, didChange newState: RTCIceGatheringState) {}
    func peerConnection(_ pc: RTCPeerConnection, didRemove candidates: [RTCIceCandidate]) {}
    func peerConnection(_ pc: RTCPeerConnection, didOpen dataChannel: RTCDataChannel) {}
}
